File: assets/src/js/lib/discover/lang.js
// Language detection for stream titles.
//
// Mirror of services/stremio/lang.go (code, name, flag, titleAliases):
// TestLangJSMirrorsTheGoTable parses this file and compares the two, so an
// entry added on one side and not the other is a red test rather than a
// surface that quietly disagrees with the server.
//
// `titleAliases` are the tokens that mean this language IN A TITLE, which
// is a different question from what the language is called. A row with
// none is nameable and choosable but never detected -- the safe default
// for tokens nobody has measured against real release vocabulary, because
// a wrong detection pre-empts the Cyrillic fallback below and, server
// side, drops the release out of an exclusive filter.
const LANGUAGES = [
    { code: 'en', name: 'English',     flag: '🇬🇧', titleAliases: ['eng', 'english', 'en'], extraFlags: ['🇺🇸', '🇦🇺'] },
    { code: 'ru', name: 'Russian',     flag: '🇷🇺', titleAliases: ['rus', 'russian', 'ru', 'рус', 'русский'] },
    { code: 'uk', name: 'Ukrainian',   flag: '🇺🇦', titleAliases: ['ukr', 'ukrainian', 'ua', 'укр', 'українська'] },
    { code: 'it', name: 'Italian',     flag: '🇮🇹', titleAliases: ['ita', 'italian', 'it'] },
    { code: 'fr', name: 'French',      flag: '🇫🇷', titleAliases: ['fre', 'french', 'fr'] },
    { code: 'es', name: 'Spanish',     flag: '🇪🇸', titleAliases: ['spa', 'spanish', 'es'] },
    { code: 'de', name: 'German',      flag: '🇩🇪', titleAliases: ['ger', 'german', 'de'] },
    { code: 'pt', name: 'Portuguese',  flag: '🇧🇷', titleAliases: ['por', 'portuguese', 'pt'], extraFlags: ['🇵🇹'] },
    { code: 'cs', name: 'Czech',       flag: '🇨🇿', titleAliases: ['cze', 'czech', 'cz'] },
    { code: 'pl', name: 'Polish',      flag: '🇵🇱', titleAliases: ['pol', 'polish', 'pl'] },
    { code: 'nl', name: 'Dutch',       flag: '🇳🇱', titleAliases: ['dut', 'dutch', 'nl'] },
    { code: 'ja', name: 'Japanese',    flag: '🇯🇵', titleAliases: ['jpn', 'japanese', 'ja'] },
    { code: 'ko', name: 'Korean',      flag: '🇰🇷', titleAliases: ['kor', 'korean', 'ko'] },
    { code: 'zh', name: 'Chinese',     flag: '🇨🇳', titleAliases: ['chi', 'chinese', 'zh'] },
    { code: 'ar', name: 'Arabic',      flag: '🇸🇦', titleAliases: ['ara', 'arabic', 'ar'] },
    { code: 'hi', name: 'Hindi',       flag: '🇮🇳', titleAliases: ['hin', 'hindi', 'hi'] },
    { code: 'tr', name: 'Turkish',     flag: '🇹🇷', titleAliases: ['tur', 'turkish', 'tr'] },
    { code: 'sv', name: 'Swedish',     flag: '🇸🇪', titleAliases: ['swe', 'swedish', 'sv'] },
    { code: 'no', name: 'Norwegian',   flag: '🇳🇴', titleAliases: ['nor', 'norwegian', 'no'] },
    { code: 'da', name: 'Danish',      flag: '🇩🇰', titleAliases: ['dan', 'danish', 'da'] },
    { code: 'fi', name: 'Finnish',     flag: '🇫🇮', titleAliases: ['fin', 'finnish', 'fi'] },
    { code: 'ro', name: 'Romanian',    flag: '🇷🇴', titleAliases: ['rum', 'romanian', 'ro'] },
    { code: 'hu', name: 'Hungarian',   flag: '🇭🇺', titleAliases: ['hun', 'hungarian', 'hu'] },
    { code: 'el', name: 'Greek',       flag: '🇬🇷', titleAliases: ['gre', 'greek', 'el'] },
    { code: 'bg', name: 'Bulgarian',   flag: '🇧🇬', titleAliases: ['bul', 'bulgarian', 'bg'] },
    { code: 'hr', name: 'Croatian',    flag: '🇭🇷', titleAliases: ['hrv', 'croatian', 'hr'] },
    { code: 'sr', name: 'Serbian',     flag: '🇷🇸', titleAliases: ['srp', 'serbian', 'sr'] },
    { code: 'sl', name: 'Slovenian',   flag: '🇸🇮', titleAliases: ['slv', 'slovenian', 'sl'] },
    { code: 'he', name: 'Hebrew',      flag: '🇮🇱', titleAliases: ['heb', 'hebrew', 'he'] },
    { code: 'th', name: 'Thai',        flag: '🇹🇭', titleAliases: ['tha', 'thai', 'th'] },
    { code: 'vi', name: 'Vietnamese',  flag: '🇻🇳', titleAliases: ['vie', 'vietnamese', 'vi'] },
    { code: 'id', name: 'Indonesian',  flag: '🇮🇩', titleAliases: ['ind', 'indonesian', 'id'] },
    { code: 'ms', name: 'Malay',       flag: '🇲🇾', titleAliases: ['may', 'malay', 'ms'] },
    // Appended 2026-09-16, mirroring services/stremio/lang.go. They carry
    // NO titleAliases, exactly as the Go rows do: they exist to be named
    // and chosen, never found in a title (see the note there -- 'KAT' is
    // KickassTorrents, not Georgian). A Go test compares this list with
    // the Go one field by field, so the two cannot drift.
    { code: 'sk', name: 'Slovak',      flag: '🇸🇰', titleAliases: [] },
    { code: 'lt', name: 'Lithuanian',  flag: '🇱🇹', titleAliases: [] },
    { code: 'lv', name: 'Latvian',     flag: '🇱🇻', titleAliases: [] },
    { code: 'et', name: 'Estonian',    flag: '🇪🇪', titleAliases: [] },
    { code: 'fa', name: 'Persian',     flag: '🇮🇷', titleAliases: [] },
    { code: 'bn', name: 'Bengali',     flag: '🇧🇩', titleAliases: [] },
    { code: 'ta', name: 'Tamil',       flag: '🇱🇰', titleAliases: [] },
    { code: 'kk', name: 'Kazakh',      flag: '🇰🇿', titleAliases: [] },
    { code: 'ka', name: 'Georgian',    flag: '🇬🇪', titleAliases: [] },
    { code: 'hy', name: 'Armenian',    flag: '🇦🇲', titleAliases: [] },
    { code: 'az', name: 'Azerbaijani', flag: '🇦🇿', titleAliases: [] },
    { code: 'ca', name: 'Catalan',     flag: '🇦🇩', titleAliases: [] },
    // Latino has no Go counterpart and no code: it is a release-title tag,
    // not a language of the settings list, so the mirror test skips it by
    // the empty `code`. Its flag is deliberately Spanish's -- see the
    // first-wins registration below, which is what keeps a bare 🇪🇸
    // resolving to Spanish.
    { code: '', name: 'Latino', flag: '🇪🇸', titleAliases: ['lat', 'latino'], extraFlags: ['🇲🇽', '🇦🇷'] },
];

// LANG_MAP resolves a title token (alias / short code / flag emoji) to a
// language. Built from titleAliases alone, so a row with none is absent
// from it, flag included -- the Go side does the same.
//
// First key wins. Latino and Spanish share 🇪🇸, and last-wins made a bare
// 🇪🇸 in a title resolve to Latino, shadowing Spanish since the row was
// added; 🇲🇽 and 🇦🇷 still reach Latino, which is the distinction that
// tag is for.
export const LANG_MAP = {};
for (const lang of LANGUAGES) {
    if (!lang.titleAliases || lang.titleAliases.length === 0) continue;
    const entry = { flag: lang.flag, name: lang.name };
    const put = (k) => { if (!(k in LANG_MAP)) LANG_MAP[k] = entry; };
    for (const a of lang.titleAliases) put(a);
    put(lang.flag);
    if (lang.extraFlags) for (const f of lang.extraFlags) put(f);
}

// Words to skip -- they are not languages even though they match short codes
const LANG_SKIP = new Set([
    'no', // Norwegian conflicts with "no" (e.g. "No torrent")
]);

// supportsFlagEmoji reports whether the platform actually renders
// regional-indicator flag emoji. Windows (every browser except Firefox,
// which ships its own emoji font) falls back to letter pairs like "RU",
// so flag-decorated chips look broken there. Detection: draw a flag on
// a canvas and look for a colored (non-grayscale) pixel — the letter
// fallback is monochrome. Result is computed once per page.
let flagEmojiSupport = null;
export function supportsFlagEmoji() {
    if (flagEmojiSupport != null) return flagEmojiSupport;
    try {
        const canvas = document.createElement('canvas');
        canvas.width = 20;
        canvas.height = 20;
        const ctx = canvas.getContext('2d', { willReadFrequently: true });
        ctx.textBaseline = 'top';
        ctx.font = '16px sans-serif';
        ctx.fillText('🇷🇺', 0, 0);
        const data = ctx.getImageData(0, 0, 20, 20).data;
        flagEmojiSupport = false;
        for (let i = 0; i < data.length; i += 4) {
            if (data[i + 3] > 0 && !(data[i] === data[i + 1] && data[i + 1] === data[i + 2])) {
                flagEmojiSupport = true;
                break;
            }
        }
    } catch {
        flagEmojiSupport = true; // can't tell — assume the common case
    }
    return flagEmojiSupport;
}

const RU_VOICE_OVER = new Set(['avo', 'mvo', 'dvo']);

export function extractLanguages(title) {
    if (!title) return [];
    const found = {};
    const tokens = title.split(/[\s./()[\],|+]+/);
    for (const t of tokens) {
        const trimmed = t.trim();
        if (!trimmed) continue;
        const lower = trimmed.toLowerCase();
        if (LANG_SKIP.has(lower)) continue;
        const lang = LANG_MAP[lower];
        if (lang && !found[lang.name]) {
            found[lang.name] = lang;
        }
        // Voice-over abbreviations only the Russian scene uses. A tracker
        // release is routinely titled in transliterated English with
        // nothing but "AVO"/"MVO"/"DVO" to say what language it is in.
        if (RU_VOICE_OVER.has(lower)) {
            const ru = LANG_MAP['ru'];
            if (ru && !found[ru.name]) found[ru.name] = ru;
        }
    }
    // Only when nothing was tagged explicitly: Cyrillic is itself the tag.
    // Ukrainian-only letters mean Ukrainian, anything else Cyrillic is
    // Russian on the trackers these titles come from. Keep in sync with
    // ExtractLanguages in services/stremio/lang.go.
    if (Object.keys(found).length === 0 && /[\u0400-\u04FF]/.test(title)) {
        const lang = LANG_MAP[/[іїєґІЇЄҐ]/.test(title) ? 'ukr' : 'ru'];
        if (lang) found[lang.name] = lang;
    }
    return Object.values(found);
}

File: services/stremio/lang.go
package stremio

import "strings"

// Language describes a language entry exposed to the Stremio settings UI
// and used for filtering stream titles. Mirrors the LANGUAGES list in
// assets/src/js/lib/discover/lang.js so the Stremio addon and the
// Discover stream modal share the same detection rules.
type Language struct {
	Code    string
	Name    string
	Flag    string
	Aliases []string
}

// Languages is the canonical, ordered list of supported languages. Keep in
// sync with assets/src/js/lib/discover/lang.js.
var Languages = []Language{
	{Code: "en", Name: "English", Flag: "🇬🇧", Aliases: []string{"eng", "english", "en"}},
	{Code: "ru", Name: "Russian", Flag: "🇷🇺", Aliases: []string{"rus", "russian", "ru", "рус", "русский"}},
	{Code: "uk", Name: "Ukrainian", Flag: "🇺🇦", Aliases: []string{"ukr", "ukrainian", "ua", "укр", "українська"}},
	{Code: "it", Name: "Italian", Flag: "🇮🇹", Aliases: []string{"ita", "italian", "it"}},
	{Code: "fr", Name: "French", Flag: "🇫🇷", Aliases: []string{"fre", "french", "fr"}},
	{Code: "es", Name: "Spanish", Flag: "🇪🇸", Aliases: []string{"spa", "spanish", "es"}},
	{Code: "de", Name: "German", Flag: "🇩🇪", Aliases: []string{"ger", "german", "de"}},
	{Code: "pt", Name: "Portuguese", Flag: "🇧🇷", Aliases: []string{"por", "portuguese", "pt"}},
	{Code: "cs", Name: "Czech", Flag: "🇨🇿", Aliases: []string{"cze", "czech", "cz"}},
	{Code: "pl", Name: "Polish", Flag: "🇵🇱", Aliases: []string{"pol", "polish", "pl"}},
	{Code: "nl", Name: "Dutch", Flag: "🇳🇱", Aliases: []string{"dut", "dutch", "nl"}},
	{Code: "ja", Name: "Japanese", Flag: "🇯🇵", Aliases: []string{"jpn", "japanese", "ja"}},
	{Code: "ko", Name: "Korean", Flag: "🇰🇷", Aliases: []string{"kor", "korean", "ko"}},
	{Code: "zh", Name: "Chinese", Flag: "🇨🇳", Aliases: []string{"chi", "chinese", "zh"}},
	{Code: "ar", Name: "Arabic", Flag: "🇸🇦", Aliases: []string{"ara", "arabic", "ar"}},
	{Code: "hi", Name: "Hindi", Flag: "🇮🇳", Aliases: []string{"hin", "hindi", "hi"}},
	{Code: "tr", Name: "Turkish", Flag: "🇹🇷", Aliases: []string{"tur", "turkish", "tr"}},
	{Code: "sv", Name: "Swedish", Flag: "🇸🇪", Aliases: []string{"swe", "swedish", "sv"}},
	{Code: "no", Name: "Norwegian", Flag: "🇳🇴", Aliases: []string{"nor", "norwegian", "no"}},
	{Code: "da", Name: "Danish", Flag: "🇩🇰", Aliases: []string{"dan", "danish", "da"}},
	{Code: "fi", Name: "Finnish", Flag: "🇫🇮", Aliases: []string{"fin", "finnish", "fi"}},
	{Code: "ro", Name: "Romanian", Flag: "🇷🇴", Aliases: []string{"rum", "romanian", "ro"}},
	{Code: "hu", Name: "Hungarian", Flag: "🇭🇺", Aliases: []string{"hun", "hungarian", "hu"}},
	{Code: "el", Name: "Greek", Flag: "🇬🇷", Aliases: []string{"gre", "greek", "el"}},
	{Code: "bg", Name: "Bulgarian", Flag: "🇧🇬", Aliases: []string{"bul", "bulgarian", "bg"}},
	{Code: "hr", Name: "Croatian", Flag: "🇭🇷", Aliases: []string{"hrv", "croatian", "hr"}},
	{Code: "sr", Name: "Serbian", Flag: "🇷🇸", Aliases: []string{"srp", "serbian", "sr"}},
	{Code: "sl", Name: "Slovenian", Flag: "🇸🇮", Aliases: []string{"slv", "slovenian", "sl"}},
	{Code: "he", Name: "Hebrew", Flag: "🇮🇱", Aliases: []string{"heb", "hebrew", "he"}},
	{Code: "th", Name: "Thai", Flag: "🇹🇭", Aliases: []string{"tha", "thai", "th"}},
	{Code: "vi", Name: "Vietnamese", Flag: "🇻🇳", Aliases: []string{"vie", "vietnamese", "vi"}},
	{Code: "id", Name: "Indonesian", Flag: "🇮🇩", Aliases: []string{"ind", "indonesian", "id"}},
	{Code: "ms", Name: "Malay", Flag: "🇲🇾", Aliases: []string{"may", "malay", "ms"}},
	// Appended 2026-09-16 so every code the subtitle-translate service
	// accepts has an entry here (langSupersetTest pins that). Order of the
	// entries above is unchanged: the Stremio settings list and the
	// language row read this order, and reshuffling it would move chips
	// under people.
	//
	// Aliases are chosen more narrowly than the rows above, because every
	// one of them is matched against whitespace-split tokens of torrent
	// titles (ExtractLanguages): the ISO 639-2/B codes "per", "arm", "ben",
	// "lit" and "cat" are ordinary English words or names and are left out,
	// and "et"/"ca" are listed but skipped below for the same reason.
	// Missing a language tag costs one filter chip; inventing one puts a
	// release in a language nobody asked for.
	{Code: "sk", Name: "Slovak", Flag: "🇸🇰", Aliases: []string{"slk", "slovak", "sk", "slovenčina"}},
	{Code: "lt", Name: "Lithuanian", Flag: "🇱🇹", Aliases: []string{"lithuanian", "lt", "lietuvių"}},
	{Code: "lv", Name: "Latvian", Flag: "🇱🇻", Aliases: []string{"lav", "latvian", "lv", "latviešu"}},
	{Code: "et", Name: "Estonian", Flag: "🇪🇪", Aliases: []string{"est", "estonian", "et", "eesti"}},
	{Code: "fa", Name: "Persian", Flag: "🇮🇷", Aliases: []string{"fas", "persian", "farsi", "fa", "فارسی"}},
	{Code: "bn", Name: "Bengali", Flag: "🇧🇩", Aliases: []string{"bengali", "bn", "বাংলা"}},
	// Sri Lanka, not India: Tamil is official in both, and 🇮🇳 is already
	// Hindi's. The flag is a map key (langMap), so a duplicate would
	// silently shadow the entry above it.
	{Code: "ta", Name: "Tamil", Flag: "🇱🇰", Aliases: []string{"tam", "tamil", "ta", "தமிழ்"}},
	{Code: "kk", Name: "Kazakh", Flag: "🇰🇿", Aliases: []string{"kaz", "kazakh", "kk", "қазақ"}},
	{Code: "ka", Name: "Georgian", Flag: "🇬🇪", Aliases: []string{"kat", "georgian", "ka", "ქართული"}},
	{Code: "hy", Name: "Armenian", Flag: "🇦🇲", Aliases: []string{"hye", "armenian", "hy", "հայերեն"}},
	{Code: "az", Name: "Azerbaijani", Flag: "🇦🇿", Aliases: []string{"aze", "azerbaijani", "az", "azərbaycan"}},
	// Andorra: the one state where Catalan is the sole official language,
	// and 🇪🇸 is already Spanish's.
	{Code: "ca", Name: "Catalan", Flag: "🇦🇩", Aliases: []string{"catalan", "ca", "català"}},
}

// langMap resolves an alias / 2-letter code / flag emoji to a Language entry.
var langMap = func() map[string]*Language {
	m := make(map[string]*Language, len(Languages)*4)
	for i := range Languages {
		l := &Languages[i]
		for _, a := range l.Aliases {
			m[a] = l
		}
		m[l.Flag] = l
	}
	return m
}()

// langSkip mirrors LANG_SKIP in lang.js — short tokens that look like
// language codes but produce too many false positives.
var langSkip = map[string]bool{
	"no": true, // Norwegian conflicts with the English word "no"
	"et": true, // Estonian conflicts with the French and Latin "et"
	"ca": true, // Catalan conflicts with "CA" the region code and "ca." circa
}

// langSplitter mirrors the JS regex /[\s./()[\],|+]+/ used to tokenise
// stream titles in extractLanguages.
func splitTitle(s string) []string {
	return strings.FieldsFunc(s, func(r rune) bool {
		switch r {
		case ' ', '\t', '\n', '\r',
			'.', '/', '(', ')', '[', ']', ',', '|', '+':
			return true
		}
		return false
	})
}

// ExtractLanguages detects language tags in a stream/torrent title using the
// same rules as Discover's extractLanguages() (assets/src/js/lib/discover/lang.js).
// Returns languages in first-seen order, deduplicated by Name.
func ExtractLanguages(title string) []*Language {
	if title == "" {
		return nil
	}
	seen := make(map[string]bool)
	var out []*Language
	for _, t := range splitTitle(title) {
		if t == "" {
			continue
		}
		lower := strings.ToLower(t)
		if langSkip[lower] {
			continue
		}
		if l, ok := langMap[lower]; ok && !seen[l.Name] {
			seen[l.Name] = true
			out = append(out, l)
		}
		// Voice-over abbreviations that only the Russian scene uses. A
		// rutracker release is routinely titled in transliterated English
		// with nothing but "AVO"/"MVO"/"DVO" to say what language it is in,
		// and a user who asked for Russian would otherwise have exactly
		// those releases filtered out.
		if ruVoiceOver[lower] {
			addLang(&out, seen, "ru")
		}
	}
	// Only when nothing was tagged explicitly: Cyrillic in the title is
	// itself the tag. Ukrainian-only letters mean Ukrainian, anything else
	// Cyrillic is Russian in practice on the trackers these titles come
	// from. Running this as a fallback rather than an addition keeps a
	// Ukrainian release from also counting as Russian.
	if len(out) == 0 && hasCyrillic(title) {
		if hasUkrainianLetters(title) {
			addLang(&out, seen, "uk")
		} else {
			addLang(&out, seen, "ru")
		}
	}
	return out
}

// ruVoiceOver are dubbing markers specific to the Russian scene: авторский,
// многоголосый and двухголосый voice-over. Deliberately excludes the bare
// "VO" and "DUB", which every scene uses.
var ruVoiceOver = map[string]bool{"avo": true, "mvo": true, "dvo": true}

func addLang(out *[]*Language, seen map[string]bool, code string) {
	l := LanguageByCode(code)
	if l == nil || seen[l.Name] {
		return
	}
	seen[l.Name] = true
	*out = append(*out, l)
}

func hasCyrillic(s string) bool {
	for _, r := range s {
		if r >= 0x0400 && r <= 0x04FF {
			return true
		}
	}
	return false
}

// hasUkrainianLetters looks for the letters Ukrainian has and Russian does
// not, in either case.
func hasUkrainianLetters(s string) bool {
	return strings.ContainsAny(s, "іїєґІЇЄҐ")
}

// LanguageByCode returns the Language entry with the given 2-letter code, or nil.
func LanguageByCode(code string) *Language {
	for i := range Languages {
		if Languages[i].Code == code {
			return &Languages[i]
		}
	}
	return nil
}
